src/compiler/checker.ts

```diff
@@ -32762,23 +32775,29 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkAssertionWorker(errNode: Node, type: TypeNode, expression: UnaryExpression | Expression, checkMode?: CheckMode) {
-        let exprType = checkExpression(expression, checkMode);
+        const exprType = checkExpression(expression, checkMode);
         if (isConstTypeReference(type)) {
             if (!isValidConstAssertionArgument(expression)) {
                 error(expression, Diagnostics.A_const_assertions_can_only_be_applied_to_references_to_enum_members_or_string_number_boolean_array_or_object_literals);
             }
             return getRegularTypeOfLiteralType(exprType);
         }
         checkSourceElement(type);
-        exprType = getRegularTypeOfObjectLiteral(getBaseTypeOfLiteralType(exprType));
         const targetType = getTypeFromTypeNode(type);
         if (!isErrorType(targetType)) {
             addLazyDiagnostic(() => {
-                const widenedType = getWidenedType(exprType);
+                const regularType = getRegularTypeOfObjectLiteral(getBaseTypeOfLiteralType(exprType));
+                const widenedType = getWidenedType(regularType);
                 if (!isTypeComparableTo(targetType, widenedType)) {
-                    checkTypeComparableTo(exprType, targetType, errNode,
+                    checkTypeComparableTo(regularType, targetType, errNode,
                         Diagnostics.Conversion_of_type_0_to_type_1_may_be_a_mistake_because_neither_type_sufficiently_overlaps_with_the_other_If_this_was_intentional_convert_the_expression_to_unknown_first);
                 }
+                // Assertions, as a side effect, disable widening - some casts may rely on this, so
+                // if the expression type was widenable, we shouldn't assume the cast is extraneous.
+                // (Generally speaking, such casts are better served with `as const` casts nowadays though!)
+                else if (widenedType === exprType && isTypeIdenticalTo(targetType, widenedType)) {
+                    errorOrSuggestion(!!compilerOptions.noUnnecessaryCasts, errNode, Diagnostics.Type_cast_has_no_effect_on_the_type_of_this_expression);

```

Comment:
This line is the core of the addition, and pretty much everything else in `checker` is to handle edge cases the eslint rule authors identified - like using a cast to prevent widening (a common occurrence before `as const`) or using a non-null assertion to prevent a use-before-assignment error on a non-nullable type'd variable.